presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java

```diff
@@ -173,40 +181,82 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
+    {
+        ScalarFunction scalarFunction = method.getAnnotation(ScalarFunction.class);
+        ScalarOperator scalarOperator = method.getAnnotation(ScalarOperator.class);
+        Description descriptionAnnotation = method.getAnnotation(Description.class);
+
+        ImmutableList.Builder<SqlScalarFunction> builder = ImmutableList.builder();
+        Map<Set<TypeParameter>, Constructor<?>> constructors = ImmutableMap.of();
+        if (!Modifier.isStatic(method.getModifiers())) {
+            try {
+                constructors = ImmutableMap.of(ImmutableSet.of(), method.getDeclaringClass().getConstructor());
+            }
+            catch (NoSuchMethodException e) {
+                throw new PrestoException(FUNCTION_IMPLEMENTATION_ERROR, format("%s is non-static, but its declaring class is missing a default constructor", method));
+            }
+        }
+
+        if (scalarFunction != null) {
+            String baseName = scalarFunction.value().isEmpty() ? camelToSnake(method.getName()) : scalarFunction.value();
+            builder.add(createReflectionParametricScalar(baseName, scalarFunction.hidden(), scalarFunction.deterministic(), descriptionAnnotation, ImmutableList.of(method), baseName, constructors));
+            for (String alias : scalarFunction.alias()) {
+                builder.add(createReflectionParametricScalar(alias, scalarFunction.hidden(), scalarFunction.deterministic(), descriptionAnnotation, ImmutableList.of(method), baseName, constructors));
+            }
         }
-        else {
-            name = mangleOperatorName(operatorAnnotation.value());
-            hidden = true;
-            deterministic = true;
+
+        if (scalarOperator != null) {
+            builder.add(createReflectionParametricScalar(mangleOperatorName(scalarOperator.value()), true, true, descriptionAnnotation, ImmutableList.of(method), method.getName(), constructors));
         }
 
+        return builder.build();
+    }
+
+    private static String camelToSnake(String name)
+    {
+        return LOWER_CAMEL.to(LOWER_UNDERSCORE, name);
+    }
+
+    private static SqlScalarFunction createReflectionParametricScalar(String name, boolean hidden, boolean deterministic, Description descriptionAnnotation, List<Method> methods, String objectName, Map<Set<TypeParameter>, Constructor<?>> constructors)
+    {
+        checkArgument(!name.isEmpty());
+        String description = descriptionAnnotation == null ? "" : descriptionAnnotation.value();
+
         ImmutableMap.Builder<Signature, Implementation> exactImplementations = ImmutableMap.builder();
         ImmutableList.Builder<Implementation> specializedImplementations = ImmutableList.builder();
         ImmutableList.Builder<Implementation> implementations = ImmutableList.builder();
         Signature signature = null;
 
-        Map<Set<TypeParameter>, Constructor<?>> constructors = findConstructors(clazz);
+        for (Method method : methods) {
+            if (name.isEmpty()) {
```

Comment:
This can never happen. You have a check that name is non-empty at the beginning of this method